kernel/sched.c

```diff
@@ -1760,61 +1760,227 @@ static uint32_t usage_now(void)
 	k_spinlock_key_t k = k_spin_lock(&usage_lock);
-	uint32_t u0 = _current_cpu->usage0, now = usage_now();
-	uint64_t ret = thread->base.usage;
 
-	if (u0 != 0) {
-		uint32_t dt = now - u0;
+	if (cpu->track_usage != 0) {
+		z_sched_usage_stop_internal(cpu);
+	}
+
+	k_spin_unlock(&usage_lock, k);
+}
 
 #ifdef CONFIG_SCHED_THREAD_USAGE_ALL
-		if (z_is_idle_thread_object(thread)) {
-			_kernel.idle_thread_usage += dt;
-		} else {
-			_kernel.all_thread_usage += dt;
-		}
+void z_sched_cpu_usage(uint8_t core_id, struct k_thread_runtime_stats *stats)
+{
+	k_spinlock_key_t  key;
+	struct _cpu *cpu;
+	uint32_t  now;
+	uint32_t  u0;
+
+	cpu = _current_cpu;
+	key = k_spin_lock(&usage_lock);
+
+	u0 = cpu->usage0;
+	now = usage_now();
+
+	if ((u0 != 0) && (&_kernel.cpus[core_id] == cpu)) {
+		uint32_t dt = now - u0;
+
+		sched_update_usage(cpu, cpu->current, dt);
+
+		cpu->usage0 = now;
+	}
+
+	stats->current_cycles   = cpu->usage.current;
+	stats->peak_cycles      = cpu->usage.peak;
+	stats->total_cycles     = cpu->usage.total;
+	if (cpu->usage.num_windows == 0) {
+		stats->average_cycles = 0;
+	} else {
+		stats->average_cycles = stats->total_cycles /
+					cpu->usage.num_windows;
+	}
+	stats->idle_cycles =
+		_kernel.cpus[core_id].idle_thread->base.usage.total;
+
+	stats->execution_cycles = stats->total_cycles + stats->idle_cycles;
+
+	k_spin_unlock(&usage_lock, key);
+}
 #endif
 
-		ret += dt;
-		thread->base.usage = ret;
-		_current_cpu->usage0 = now;
+void z_sched_thread_usage(struct k_thread *thread,
+			  struct k_thread_runtime_stats *stats)
+{
+	uint32_t u0;
+	uint32_t now;
+	struct _cpu *cpu;
+	k_spinlock_key_t key;
+
+	cpu = _current_cpu;
+	key = k_spin_lock(&usage_lock);
+
+	u0 = cpu->usage0;
+	now = usage_now();
+
+	if ((u0 != 0) && (thread == cpu->current)) {
+		uint32_t dt = now - u0;
+
+		/*
+		 * Update the thread's usage stats if it is the current thread
+		 * running on the current core. This restriction exsits because
+		 * it is not reliable to update the thread's usage stats if it
+		 * is running on another core. (Not only could the other core
+		 * be referencing a different clock, but even if it were using
+		 * the same clock, the various core's [usage0] fields differ.)
+		 */
+
+		sched_update_usage(cpu, thread, dt);
+
+		cpu->usage0 = now;
 	}
 
-	k_spin_unlock(&usage_lock, k);
-	return ret;
+	/* Copy-out the thread's usage stats */
+
+	stats->current_cycles   = thread->base.usage.current;
+	stats->peak_cycles      = thread->base.usage.peak;
+	stats->total_cycles     = thread->base.usage.total;
+	if (thread->base.usage.num_windows == 0) {
+		stats->average_cycles = 0;
+	} else {
+		stats->average_cycles = stats->total_cycles /
+					thread->base.usage.num_windows;
+	}
+	stats->idle_cycles     = 0;
+
+	stats->execution_cycles = stats->total_cycles;
+
+	k_spin_unlock(&usage_lock, key);
 }
 
+void k_runtime_stats_enable(void)
+{
+	k_spinlock_key_t  key;
+
+	key = k_spin_lock(&usage_lock);
+
+	if (_current_cpu->track_usage == 0) {
```

Comment:
Move this test (and the other symmetric relatives) out of the lock.  The track_usage byte is only set from the current CPU (meaning it can't race against other contexts unless our local CPU gets preempted) and within a spinlock (meaning we can't be preempted).

Not having to take and release a lock on every context switch is actually big win.

Reply:
Sorry, to clarify: not having to take and release a lock twice on every context switch and once on every ISR is actually a big win. :)

Reply:
You're right. I should've seen that.

Reply:
Wait a minute, that claim isn't right.

There can be context switch between the test and the when the lock is taken. This can result in another thread being scheduled, which could in turn attempt to do another 'enable'.  Most of what is going to happen in this particular if-block will be inconsequential if it gets executed twice. However, the num_windows++ can get executed multiple times  resulting in a lower average calculation.

Is the resulting error in the average calculation going to be significant? Early on, perhaps. But as the system continues to execute the error can be expected to assymptotically reduce to zero.

The purist in me wants to keep the error to zero. However, the pragmatic side also recognizes that it *might* be worth the trade-off. Thoughts?

Reply:
Self-forehead slap ... atomic test-and-set/test-and-clear.

Reply:
Hmmm ... Even atomics do not quite work for what is needed.